### PR TITLE
Try to make requests during 180 seconds if API is in maintenance

### PR DESCRIPTION
**Pull request by brmzkw**

### CODE REVIEW

**Comment by kdeldycke** [comment]
:+1:

**Comment by moul** [comment]
LGTM

**Comment by EvaSDK** [comment]
Note that while this seems like a nice feature, this contradicts what you would expect from a client, reading the RFC:

"... the client SHOULD handle the response as it would for a 500 response."
http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

You should imho retry more server errors (like 502 and 504 that are likely to be temporary) in this case or retry none. Implementing Retry-After header support would deal with what you intend to do in a RFC compliant way.

**Comment by bchatelard** [comment]
LGTM

**Comment by brmzkw** [comment]
Ideally, we could parse the `Retry-After` header if it exists to sleep during an amount of time expected by the server.

Parsing the `Retry-After` header is more complex than it appears, since according the RFC it can be a [HTTP-date or a delta in seconds](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html), the HTTP-date being a [`rfc1123-date | rfc850-date | asctime-date`](http://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.3.1).

A "good" (but not perfect) way to parse the header would be something like:

```python
from email.utils import parsedate
from datetime import datetime, timedelta

date = ...
next_retry = None
try:
   next_retry = datetime.now() + timedelta(seconds=int(date))
except ValueError:
  # some error handling required here
  in_secs = (datetime(*parsedate(date)[:6]) - datetime.now()).total_seconds()
  if in_secs <= 0:
    in_secs = 3 # or whatever default seems reasonable
  next_retry = datetime.now() + timedelta(seconds=in_secs)
```

In short, dealing with the `Retry-After` header is complex and I prefer to ignore it and use reasonable hardcoded values (3 seconds between two retries, maximum 60 retries).

Anyway, I also retry on 502 and 504 as you suggested.
